File: OneQuestionPerDay/300/longest_increasing_subsequence2.go
package main

func lengthOfLIS2(nums []int) int {
	tail := make([]int, 0)
	for i := 0; i < len(nums); i++ {
		if i == 0 || nums[i] > tail[len(tail)-1] {
			tail = append(tail, nums[i])
		} else {
			left, right := 0, len(tail)
			for left < right {
				mid := (left + right) / 2
				if tail[mid] < nums[i] {
					left = mid + 1
				} else {
					right = mid
				}
			}
			tail[left] = nums[i]
		}
	}
	return len(tail)
}

File: OneQuestionPerDay/300/longest_increasing_subsequence2.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tail = []
        for i in range(len(nums)):
            if i == 0 or nums[i] > tail[-1]:
                tail.append(nums[i])
            else:
                l, r = 0, len(tail)
                while l < r:
                    m = (l + r) // 2
                    if tail[m] < nums[i]:
                        l = m + 1
                    else:
                        r = m
                tail[l] = nums[i]
        return len(tail)
